components/style/gecko/conversions.rs

```diff
@@ -338,6 +340,182 @@ impl nsStyleImage {
+                            LengthOrPercentage::Percentage(Percentage(0.0)) => Y::Top,
+                            LengthOrPercentage::Percentage(Percentage(1.0)) => Y::Bottom,
+                            x => panic!("Found unexpected mBgPosY in nsStyleImage: {:?}", x),
+                        };
+                        LineDirection::Corner(horizontal_direction, vertical_direction)
+                    },
+                    _ => {
+                        LineDirection::Angle(Angle::from_gecko_style_coord(&gecko_gradient.mAngle)
+                                             .expect("gecko_gradient.mAngle has valid data"))
+                    }
+                };
+                GradientKind::Linear(line_direction)
+            },
+            _ => {
+                let gecko_size_to_keyword = |gecko_size| {
+                    match gecko_size {
+                        NS_STYLE_GRADIENT_SIZE_CLOSEST_SIDE => ShapeExtent::ClosestSide,
+                        NS_STYLE_GRADIENT_SIZE_FARTHEST_SIDE => ShapeExtent::FarthestSide,
+                        NS_STYLE_GRADIENT_SIZE_CLOSEST_CORNER => ShapeExtent::ClosestCorner,
+                        NS_STYLE_GRADIENT_SIZE_FARTHEST_CORNER => ShapeExtent::FarthestCorner,
+                        // We don't choose ShapeExtent::Contain and ShapeExtent::Cover
+                        // since those 'contain' and 'cover' keywords are handled as synonyms of the standard
+                        // 'farthest-corner' and 'closest-side' respectively.
+                        x => panic!("Found unexpected gecko_size: {:?}", x),
+                    }
+                };
+
+                let shape = match gecko_gradient.mShape as u32 {
+                    NS_STYLE_GRADIENT_SHAPE_CIRCULAR => {
+                        let circle = match gecko_gradient.mSize as u32 {
+                            NS_STYLE_GRADIENT_SIZE_EXPLICIT_SIZE => {
+                                let radius = Length::from_gecko_style_coord(&gecko_gradient.mRadiusX)
+                                    .expect("mRadiusX has valid data");
+                                debug_assert!(radius ==
+                                              Length::from_gecko_style_coord(&gecko_gradient.mRadiusY).unwrap());
+                                Circle::Radius(radius)
+                            },
+                            x => Circle::Extent(gecko_size_to_keyword(x))
+                        };
+                        EndingShape::Circle(circle)
+                    },
+                    NS_STYLE_GRADIENT_SHAPE_ELLIPTICAL => {
+                        let length_percentage_keyword = match gecko_gradient.mSize as u32 {
+                            NS_STYLE_GRADIENT_SIZE_EXPLICIT_SIZE => {
+                                Ellipse::Radii(
+                                    LengthOrPercentage::from_gecko_style_coord(&gecko_gradient.mRadiusX)
+                                        .expect("mRadiusX has valid data"),
```

Comment:
Nit: AFAIK `expect` should be given the description of the error, not the description of what valid things are expected.